File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size(), product = 1;
        vector<int> result (n, 1);
        for(int i=1; i< n; i++) {
            int x = product * nums[i-1];
            result[i] = x;
            product = x; 
        }
        product = 1;
        for(int i = n-1; i>=0; i--) {
            result[i] *= product;
            product*= nums[i];
        }
        return result;
    }
};

File: Data Structures & Algorithms/sliding-window-maximum/submission-1.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int l) {
        int n = nums.length;
        ArrayList<Integer> result = new ArrayList<Integer>();
        int j = 0;
        int max = Integer.MIN_VALUE;
        while(j < l) {
            max = Math.max(max, nums[j]);
            j++;
        }
        int i = 0; j--;
        result.add(max);
        System.out.println(result);
        System.out.println("in loop");
        while(j < n - 1) {
            int incoming = nums[j+1];
            int outgoing = nums[i];
            if(incoming > max) {
                max = incoming;
            }
            else if(outgoing == max) {
                int localMax = Integer.MIN_VALUE;
                for(int k = i + 1; k <= j+1;k++){
                    localMax = Math.max(localMax, nums[k]);
                }   
                max = localMax;
            }
            result.add(max);
                j++;
                i++;
            System.out.println(String.format("adding : %d", max));
            System.out.println(result);
        }
        int len = result.size();
        int[] arr = new int[len];
        for(i = 0; i < len; i++) {
            arr[i] = result.get(i);
        }
        return arr;
    }
}
